pkg/controller/applicationlayer/applicationlayer_controller.go

```diff
@@ -0,0 +1,310 @@
+		r.isLogsCollectionEnabled(lcSpec), lcSpec.LogIntervalSeconds, lcSpec.LogRequestsPerInterval)
+
+	ch := utils.NewComponentHandler(log, r.client, r.scheme, applicationLayer)
+
+	if err = imageset.ApplyImageSet(ctx, r.client, variant, component); err != nil {
+		reqLogger.Error(err, "Error with images from ImageSet")
+		r.status.SetDegraded("Error with images from ImageSet", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	// TODO: when there are more ApplicationLayer options then it will need to be restructured, as each of the
+	// different features will not have their own CreateOrUpdateOrDelete
+	if err = ch.CreateOrUpdateOrDelete(ctx, component, r.status); err != nil {
+		reqLogger.Error(err, "Error creating / updating resource")
+		r.status.SetDegraded("Error creating / updating resource", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	// Clear the degraded bit if we've reached this far.
+	r.status.ClearDegraded()
+
+	if !r.status.IsAvailable() {
+		// Schedule a kick to check again in the near future, hopefully by then things will be available.
+		return reconcile.Result{RequeueAfter: 30 * time.Second}, nil
+	}
+
+	// Everything is available - update the CRD status.
+	applicationLayer.Status.State = operatorv1.TigeraStatusReady
+	if err = r.client.Status().Update(ctx, applicationLayer); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+// updateApplicationLayerWithDefaults populates the applicationlayer with defaults.
+func updateApplicationLayerWithDefaults(al *operatorv1.ApplicationLayer) {
+	defaultLogIntervalSeconds := int64(5)
+	defaultLogRequestsPerInterval := int64(-1)
+
+	if al.Spec.LogCollection != nil {
+		if al.Spec.LogCollection.LogRequestsPerInterval == nil {
+			al.Spec.LogCollection.LogRequestsPerInterval = &defaultLogRequestsPerInterval
+		}
+		if al.Spec.LogCollection.LogIntervalSeconds == nil {
+			al.Spec.LogCollection.LogIntervalSeconds = &defaultLogIntervalSeconds
+		}
+	}
+}
+
+// getApplicationLayer returns the default ApplicationLayer instance with defaults populated.
+func getApplicationLayer(ctx context.Context, cli client.Client) (*operatorv1.ApplicationLayer, error) {
+	instance := &operatorv1.ApplicationLayer{}
+	err := cli.Get(ctx, utils.DefaultTSEEInstanceKey, instance)
+	if err != nil {
+		return nil, err
+	}
+
+	return instance, nil
+}
+
+func (r *ReconcileApplicationLayer) isLogsCollectionEnabled(l7Spec *operatorv1.LogCollectionSpec) bool {
+	return l7Spec != nil && l7Spec.CollectLogs != nil && *l7Spec.CollectLogs == operatorv1.L7LogCollectionEnabled
+}
+
+// patchFelixTproxyMode takes all application layer specs as arguments and patches felix config.
+// If at least one of the specs requires TPROXYMode as "Enabled" it'll be patched as "Enabled" otherwise it is "Disabled".
+func (r *ReconcileApplicationLayer) patchFelixTproxyMode(ctx context.Context, l7Spec *operatorv1.LogCollectionSpec) error {
```

Comment:
I'm guessing this was discussed at a design level with Casey and Erik (so take my comments with a grain of salt), but I'm a little wary about patching FelixConfiguration. 

Is TPROXYMode something the user is allowed to have control of, or are we trying to take control away from them? If the user is allowed to have control over this, I would say we should set the status to degraded and fail if it's not set properly, and not subtly change this.

If we don't want the user to have any control over this setting then this is probably the right thing to do, although I can't really remember if we wanted non user configuration in this CR (I think we did so that the felix pod wouldn't restart on config changes, but there's a blurry line around this stuff).

Reply:
Users is not allowed to have control over `TPROXYMode`, so we are explicitly turning it on and off. User will still have control over turning the L7 logs on and off using the Application layer CRD (which in turn, turns this mode on and off along with creating/deleting the envoy daemon set).